### PR TITLE
Javascript logic

### PR DESCRIPTION
**Pull request by aannie-y**
Able to fetch response from Servlet and parse the JSON into the `callCorrespondingFunction` function which calls the corresponding function according to the error code. Also changed the Closure templating format to a table format to allow for better readability and added an additional template for `invalid_row_length`. Tests have also been implemented for for the `unknown_column` and `invalid_row_length` notices, notices not implemented and the `callCorrespondingFunction` to see if correct functions have been called.

Also removed a `response.write` line from RunValidatorServlet as it was not a JSON output and removed corresponding test lines from RunValidatorTest.java.

### CODE REVIEW

File: webapp/src/main/webapp/resources/output.js [hunk 1]
```diff
@@ -12,18 +12,32 @@
...
+    // Notice has not been implemented, output raw json
+    if (runFunction(notice.code, notice) === -1) {
+      const unknownNotice = document.createElement('p');
+      unknownNotice.innerHTML = 'NOTICE CONTAINER NOT IMPLEMENTED - RAW JSON DATA: <br><br>' + JSON.stringify(notice);
+      document.getElementById("noticeContainer").appendChild(unknownNotice);
+    };
   }
+}
+
+function runFunction(name, arguments) {
```
Comment (jupi3200) [inline webapp/src/main/webapp/resources/output.js]: perhaps a more readable name would be _runFunctionByName_

File: webapp/src/main/webapp/resources/output.js [hunk 2]
```diff
@@ -12,18 +12,32 @@
...
+  const fn = window[name];
+  // check if fn is a function
+  if (typeof fn !== 'function') return -1;
```
Comment (jupi3200) [inline webapp/src/main/webapp/resources/output.js]: If the function returns a value in some cases, if should probably heve well defined semantics and return a value in any case.

How about having this function return a boolean indicating if the notice was handled?

File: webapp/src/main/webapp/resources/output.js [hunk 3]
```diff
@@ -12,18 +12,32 @@
...
-      }
-    ]
+function callCorrespondingFunction(noticesJSON) {
+  const noticeContainer = JSON.parse(noticesJSON);
+  for (var i = 0; i < noticeContainer.notices.length; i++) {
+    const notice = noticeContainer.notices[i];
+    // Notice has not been implemented, output raw json
+    if (runFunction(notice.code, notice) === -1) {
+      const unknownNotice = document.createElement('p');
+      unknownNotice.innerHTML = 'NOTICE CONTAINER NOT IMPLEMENTED - RAW JSON DATA: <br><br>' + JSON.stringify(notice);
```
Comment (jupi3200) [inline webapp/src/main/webapp/resources/output.js]: Do we really want the prefix 'NOTICE CONTAINER NOT IMPLEMENTED - RAW JSON DATA:' repeated?

I would suggest that we have a list at the bottom of the page of notices not yet implemented, and adding the json to the list as needed.

File: webapp/src/main/webapp/resources/output.js [hunk 4]
```diff
@@ -12,18 +12,32 @@
...
+  const fn = window[name];
+  // check if fn is a function
+  if (typeof fn !== 'function') return -1;
+  fn.apply(window, [arguments]);
+}
+
+function unknown_column(params) {
   const template = goog.soy.renderAsElement(validator.templates.unknownColumnNotice, params);
-  document.getElementById("warning").appendChild(template);
+  document.getElementById("noticeContainer").appendChild(template);
```
Comment (jupi3200) [inline webapp/src/main/webapp/resources/output.js]: I thought we wanted one contained to warnings and another for errors.

File: webapp/src/main/webapp/script.js [hunk 5]
```diff
@@ -11,3 +11,26 @@
...
+    }).then((resp) => {
+      // Check if output is in JSON format. If not, error has occured.
+      const contentType = resp.headers.get("content-type");
+      if (contentType && contentType.indexOf("application/json") !== -1) {
+        return resp.text().then(data => {
+          this.callCorrespondingFunction(data);
+        });
+      } else {
+        // Upload Error
+        this.alert(resp.text());
```
Comment (jupi3200) [inline webapp/src/main/webapp/script.js]: Do we really want a popup on the website?
How about just outputing the text in a red container?

File: webapp/src/main/webapp/test/output.test.js [hunk 6]
```diff
@@ -18,9 +18,114 @@ describe('Output', function() {
...
+    const output = "<div><p class=\"error\">Error - Invalid csv row length!</p>\
+<p>Description: A row in the input file has a different number of values than specified by the CSV header.</p>\
+<p><b>1</b> Invalid row length found in:</p><table><thead><tr><th>Filename</th><th>CSV Row Number</th>\
+<th>Row length</th><th>Header count</th></tr></thead><tbody><tr><td>stop_times.txt</td>\
+<td>17</td><td>5</td><td>9</td></tr></tbody></table><p>Please set the row length as specified by the CSV header!</p>\
+<br><br></div><div><p class=\"warning\">Warning - Unknown Column(s) found!</p>\
+<p>Description: A column name is unknown.</p>\<p><b>1</b> unknown column(s) found in:</p>\
+<table><thead><tr><th>Filename</th><th>Field name</th><th>Index</th></tr></thead><tbody>\
+<tr><td>stop_times.txt</td><td>drop_off_time</td><td>8</td></tr>\
+</tbody></table><p>Please delete or rename column!</p><br><br></div>"
```
Comment (jupi3200) [inline webapp/src/main/webapp/test/output.test.js]: Could we use [Template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) to make the test more readable?

**Review by jupi3200 — COMMENTED** [review]
Can we possibly move the changes to UploadServlet and associated test to a different PR?

File: webapp/src/main/webapp/resources/output.js (hunk 1 above)
Comment (aannie-y) [reply webapp/src/main/webapp/resources/output.js]: yep ok

Comment (aannie-y) [reply webapp/src/main/webapp/resources/output.js]: Done

File: webapp/src/main/webapp/resources/output.js (hunk 3 above)
Comment (aannie-y) [reply webapp/src/main/webapp/resources/output.js]: Hmm, yes ok. Will implement this.

File: webapp/src/main/webapp/test/output.test.js (hunk 6 above)
Comment (aannie-y) [reply webapp/src/main/webapp/test/output.test.js]: I tried before but I get a syntax error for the backtick.

File: webapp/src/main/webapp/resources/output.js (hunk 2 above)
Comment (aannie-y) [reply webapp/src/main/webapp/resources/output.js]: Done!

File: webapp/src/main/webapp/resources/output.js (hunk 4 above)
Comment (aannie-y) [reply webapp/src/main/webapp/resources/output.js]: Ah yes, fixed.

File: webapp/src/main/webapp/script.js (hunk 5 above)
Comment (aannie-y) [reply webapp/src/main/webapp/script.js]: Yep ok.

File: webapp/src/main/webapp/resources/output.js (hunk 2 above)
Comment (jupi3200) [reply webapp/src/main/webapp/resources/output.js]: Is there a reason we return an integer? Why not simple true/false ?

File: webapp/src/main/webapp/test/output.test.js [hunk 7]
```diff
@@ -18,9 +19,116 @@ describe('Output', function() {
...
+        }
+      ]
+    };
+    invalid_row_length(params);
+    const output = "<div><p class=\"error\">Error - Invalid csv row length!</p>\
+<p>Description: A row in the input file has a different number of values than specified by the CSV header.</p>\
+<p><b>2</b> Invalid row length found in:</p><table><thead><tr><th>Filename</th><th>CSV Row Number</th>\
+<th>Row length</th><th>Header count</th></tr></thead><tbody><tr><td>stop_times.txt</td>\
+<td>17</td><td>5</td><td>9</td></tr><tr><td>stop_times.txt</td><td>18</td><td>5</td>\
+<td>9</td></tr></tbody></table><p>Please set the row length as specified by the CSV header!</p><br><br></div>"
```
Comment (jupi3200) [inline webapp/src/main/webapp/test/output.test.js]: Can you reformat this lines to make it more readable (Single line per row)?

File: webapp/src/main/webapp/resources/output.js (hunk 2 above)
Comment (aannie-y) [reply webapp/src/main/webapp/resources/output.js]: Fixed!

File: webapp/src/main/webapp/script.js (hunk 5 above)
Comment (aannie-y) [reply webapp/src/main/webapp/script.js]: Fixed. Error now outputs in the DOM. CSS styling to be added later on to make it a red container.

File: webapp/src/main/webapp/test/output.test.js (hunk 7 above)
Comment (aannie-y) [reply webapp/src/main/webapp/test/output.test.js]: Yep ok

Comment (aannie-y) [reply webapp/src/main/webapp/test/output.test.js]: Done. I've reformatted the lines so be a lot more readable.

File: webapp/src/main/webapp/resources/output.js [hunk 8]
```diff
@@ -12,18 +12,37 @@
...
+  const noticeContainer = JSON.parse(noticesJSON);
+  for (var i = 0; i < noticeContainer.notices.length; i++) {
+    const notice = noticeContainer.notices[i];
+    // Notice has not been implemented, output raw json
+    if (!runFunctionName(notice.code, notice)) {
+      const unknownNotice = document.createElement('p');
+      unknownNotice.innerHTML = 'NOTICE CONTAINER NOT IMPLEMENTED - RAW JSON DATA: <br><br>' + JSON.stringify(notice);
+      document.getElementById("unimplementedNotices").appendChild(unknownNotice);
+    } else {
+      console.log("Notice has been handled!");
```
Comment (jupi3200) [inline webapp/src/main/webapp/resources/output.js]: Can we avoid writing to console during non error execution?

File: webapp/src/main/webapp/resources/output.js (hunk 3 above)
Comment (aannie-y) [reply webapp/src/main/webapp/resources/output.js]: Fixed!

File: webapp/src/main/webapp/resources/output.js (hunk 8 above)
Comment (aannie-y) [reply webapp/src/main/webapp/resources/output.js]: Yep, will remove this.

Comment (aannie-y) [reply webapp/src/main/webapp/resources/output.js]: Done!

**Review by jupi3200 — APPROVED** [review]

**Comment by aannie-y** [comment]
> Can we possibly move the changes to UploadServlet and associated test to a different PR?

done

**Pull request merged**